osu.Game/OsuGame.cs

```diff
@@ -1581,12 +1581,20 @@ private void screenChanged(IScreen current, IScreen newScreen)
 
             if (current is IOsuScreen currentOsuScreen)
             {
+                if (currentOsuScreen.AllowBackButton)
+                    BackButton.State.UnbindFrom(currentOsuScreen.BackButtonState);
+
                 OverlayActivationMode.UnbindFrom(currentOsuScreen.OverlayActivationMode);
                 API.Activity.UnbindFrom(currentOsuScreen.Activity);
             }
 
             if (newScreen is IOsuScreen newOsuScreen)
             {
+                if (newOsuScreen.AllowBackButton)
+                    ((IBindable<Visibility>)BackButton.State).BindTo(newOsuScreen.BackButtonState);
+                else
+                    BackButton.Hide();
```

Comment:
Why do `AllowBackButton` and `BackButtonState` coexist? So now you have to _both_ `AllowBackButton` and also set a valid `BackButtonState`? I guess you made up for it by setting `BackButtonState` to visible but like... why have 2 virtuals for essentially the same thing?

Can we not have just one?

Reply:
Read OP.

Reply:
Is "exiting being controlled by `AllowBackButton`" even documented anywhere? It's such a weird dependency.

I dunno these virtuals are a mess. I'd like to not have such redundancy in virtual a wherein you have to coordinate two to make them work. It's just not good quality.

Reply:
I wouldn't view it that way.

`AllowBackButton` controls whether the screen can be exited, `BackButtonState` controls the state of the back button **if** `AllowBackButton` is enabled. I don't see any reason to complicate matters further IMO.

Reply:
Can we follow existing patterns like this?

https://github.com/ppy/osu/blob/03220598b8734cb93d53d15cc054d89e2d4f04ec/osu.Game/Screens/OsuScreen.cs#L50-L55

Or is it not feasible in this context for some reason?

Reply:
In relation to the conversation above, I don't see how that pattern has anything to do with what the above is discussing.

Reply:
```diff
diff --git a/osu.Game/OsuGame.cs b/osu.Game/OsuGame.cs
index dce24c6ee7..dcad91a805 100644
--- a/osu.Game/OsuGame.cs
+++ b/osu.Game/OsuGame.cs
@@ -175,6 +175,8 @@ public partial class OsuGame : OsuGameBase, IKeyBindingHandler<GlobalAction>, IL
         /// </summary>
         public readonly IBindable<OverlayActivation> OverlayActivationMode = new Bindable<OverlayActivation>();
 
+        public readonly IBindable<bool> BackButtonVisibility = new Bindable<bool>();
+
         IBindable<LocalUserPlayingState> ILocalUserPlayInfo.PlayingState => playingState;
 
         private readonly Bindable<LocalUserPlayingState> playingState = new Bindable<LocalUserPlayingState>();
@@ -1019,7 +1021,7 @@ protected override void LoadComplete()
                                             if (!(ScreenStack.CurrentScreen is IOsuScreen currentScreen))
                                                 return;
 
-                                            if (!((Drawable)currentScreen).IsLoaded || (currentScreen.AllowBackButton && !currentScreen.OnBackButton()))
+                                            if (!((Drawable)currentScreen).IsLoaded || (currentScreen.AllowExit && !currentScreen.OnBackButton()))
                                                 ScreenStack.Exit();
                                         }
                                     },
@@ -1189,6 +1191,16 @@ protected override void LoadComplete()
                 if (mode.NewValue != OverlayActivation.All) CloseAllOverlays();
             };
 
+            BackButtonVisibility.ValueChanged += visible =>
+            {
+                if (visible.NewValue)
+                    BackButton.Show();
+                else
+                {
+                    BackButton.Hide();
+                }
+            };
+
             // Importantly, this should be run after binding PostNotification to the import handlers so they can present the import after game startup.
             handleStartupImport();
         }
@@ -1582,12 +1594,16 @@ private void screenChanged(IScreen current, IScreen newScreen)
             if (current is IOsuScreen currentOsuScreen)
             {
                 OverlayActivationMode.UnbindFrom(currentOsuScreen.OverlayActivationMode);
+                BackButtonVisibility.UnbindFrom(currentOsuScreen.BackButtonVisibility);
+
                 API.Activity.UnbindFrom(currentOsuScreen.Activity);
             }
 
             if (newScreen is IOsuScreen newOsuScreen)
             {
                 OverlayActivationMode.BindTo(newOsuScreen.OverlayActivationMode);
+                BackButtonVisibility.BindTo(newOsuScreen.BackButtonVisibility);
+
                 API.Activity.BindTo(newOsuScreen.Activity);
 
                 GlobalCursorDisplay.MenuCursor.HideCursorOnNonMouseInput = newOsuScreen.HideMenuCursorOnNonMouseInput;
@@ -1597,11 +1613,6 @@ private void screenChanged(IScreen current, IScreen newScreen)
                 else
                     Toolbar.Show();
 
-                if (newOsuScreen.AllowBackButton)
-                    BackButton.Show();
-                else
-                    BackButton.Hide();
-
                 if (newOsuScreen.ShowFooter)
                 {
                     BackButton.Hide();
diff --git a/osu.Game/Overlays/Settings/Sections/Maintenance/MigrationRunScreen.cs b/osu.Game/Overlays/Settings/Sections/Maintenance/MigrationRunScreen.cs
index 3bba480aaa..ef653cce00 100644
--- a/osu.Game/Overlays/Settings/Sections/Maintenance/MigrationRunScreen.cs
+++ b/osu.Game/Overlays/Settings/Sections/Maintenance/MigrationRunScreen.cs
@@ -26,7 +26,7 @@ public partial class MigrationRunScreen : OsuScreen
         [Resolved(canBeNull: true)]
         private OsuGame game { get; set; }
 
-        public override bool AllowBackButton => false;
+        public override bool AllowExit => false;
 
         public override bool AllowExternalScreenChange => false;
 
diff --git a/osu.Game/Screens/Edit/Editor.cs b/osu.Game/Screens/Edit/Editor.cs
index 644e1afb3b..23e018151f 100644
--- a/osu.Game/Screens/Edit/Editor.cs
+++ b/osu.Game/Screens/Edit/Editor.cs
@@ -80,7 +80,7 @@ public partial class Editor : ScreenWithBeatmapBackground, IKeyBindingHandler<Gl
 
         public override float BackgroundParallaxAmount => 0.1f;
 
-        public override bool AllowBackButton => false;
+        public override bool AllowExit => false;
 
         public override bool HideOverlaysOnEnter => true;
 
diff --git a/osu.Game/Screens/Edit/EditorLoader.cs b/osu.Game/Screens/Edit/EditorLoader.cs
index 0e0fb9f795..838accc9aa 100644
--- a/osu.Game/Screens/Edit/EditorLoader.cs
+++ b/osu.Game/Screens/Edit/EditorLoader.cs
@@ -36,7 +36,7 @@ public partial class EditorLoader : ScreenWithBeatmapBackground
 
         public override float BackgroundParallaxAmount => 0.1f;
 
-        public override bool AllowBackButton => false;
+        public override bool AllowExit => false;
 
         public override bool HideOverlaysOnEnter => true;
 
diff --git a/osu.Game/Screens/IOsuScreen.cs b/osu.Game/Screens/IOsuScreen.cs
index b80c1f87a4..de99c3e83d 100644
--- a/osu.Game/Screens/IOsuScreen.cs
+++ b/osu.Game/Screens/IOsuScreen.cs
@@ -23,13 +23,13 @@ public interface IOsuScreen : IScreen
         /// <summary>
         /// Whether the user can exit this <see cref="IOsuScreen"/> by pressing the back button.
         /// </summary>
-        bool AllowBackButton { get; }
+        bool AllowExit { get; }
 
         /// <summary>
         /// Whether a footer (and a back button) should be displayed underneath the screen.
         /// </summary>
         /// <remarks>
-        /// Temporarily, the back button is shown regardless of whether <see cref="AllowBackButton"/> is true.
+        /// Temporarily, the back button is shown regardless of whether <see cref="AllowExit"/> is true.
         /// </remarks>
         bool ShowFooter { get; }
 
@@ -59,6 +59,11 @@ public interface IOsuScreen : IScreen
         /// </summary>
         IBindable<OverlayActivation> OverlayActivationMode { get; }
 
+        /// <summary>
+        /// Whether the back button should be displayed.
+        /// </summary>
+        IBindable<bool> BackButtonVisibility { get; }
+
         /// <summary>
         /// The current <see cref="UserActivity"/> for this screen.
         /// </summary>
diff --git a/osu.Game/Screens/Menu/MainMenu.cs b/osu.Game/Screens/Menu/MainMenu.cs
index 35c6bab81b..a5c81c2d38 100644
--- a/osu.Game/Screens/Menu/MainMenu.cs
+++ b/osu.Game/Screens/Menu/MainMenu.cs
@@ -48,7 +48,7 @@ public partial class MainMenu : OsuScreen, IHandlePresentBeatmap, IKeyBindingHan
 
         public override bool HideOverlaysOnEnter => Buttons == null || Buttons.State == ButtonSystemState.Initial;
 
-        public override bool AllowBackButton => false;
+        public override bool AllowExit => false;
 
         public override bool AllowExternalScreenChange => true;
 
diff --git a/osu.Game/Screens/OnlinePlay/OnlinePlayScreen.cs b/osu.Game/Screens/OnlinePlay/OnlinePlayScreen.cs
index cc6a4e09e1..2fb6f3b79d 100644
--- a/osu.Game/Screens/OnlinePlay/OnlinePlayScreen.cs
+++ b/osu.Game/Screens/OnlinePlay/OnlinePlayScreen.cs
@@ -180,7 +180,7 @@ public override bool OnBackButton()
             if (!(screenStack.CurrentScreen is IOnlinePlaySubScreen onlineSubScreen))
                 return false;
 
-            if (((Drawable)onlineSubScreen).IsLoaded && onlineSubScreen.AllowBackButton && onlineSubScreen.OnBackButton())
+            if (((Drawable)onlineSubScreen).IsLoaded && onlineSubScreen.AllowExit && onlineSubScreen.OnBackButton())
                 return true;
 
             if (screenStack.CurrentScreen != null && !(screenStack.CurrentScreen is LoungeSubScreen))
diff --git a/osu.Game/Screens/OsuScreen.cs b/osu.Game/Screens/OsuScreen.cs
index 695a074907..a3821c4153 100644
--- a/osu.Game/Screens/OsuScreen.cs
+++ b/osu.Game/Screens/OsuScreen.cs
@@ -37,7 +37,7 @@ public abstract partial class OsuScreen : Screen, IOsuScreen, IHasDescription
 
         public string Description => Title;
 
-        public virtual bool AllowBackButton => true;
+        public virtual bool AllowExit => true;
 
         public virtual bool ShowFooter => false;
 
@@ -47,6 +47,13 @@ public abstract partial class OsuScreen : Screen, IOsuScreen, IHasDescription
 
         public virtual bool HideMenuCursorOnNonMouseInput => false;
 
+        /// <summary>
+        /// The initial overlay activation mode to use when this screen is entered for the first time.
+        /// </summary>
+        protected virtual bool InitialBackButtonVisibility => false;
+
+        public readonly Bindable<bool> BackButtonVisibility;
+
         /// <summary>
         /// The initial overlay activation mode to use when this screen is entered for the first time.
         /// </summary>
@@ -154,6 +161,7 @@ protected OsuScreen()
             Origin = Anchor.Centre;
 
             OverlayActivationMode = new Bindable<OverlayActivation>(InitialOverlayActivationMode);
+            BackButtonVisibility = new BindableBool(InitialBackButtonVisibility);
         }
 
         [BackgroundDependencyLoader(true)]
diff --git a/osu.Game/Screens/Play/Player.cs b/osu.Game/Screens/Play/Player.cs
index 2d1f602832..f9239e5d97 100644
--- a/osu.Game/Screens/Play/Player.cs
+++ b/osu.Game/Screens/Play/Player.cs
@@ -57,7 +57,7 @@ public abstract partial class Player : ScreenWithBeatmapBackground, ISamplePlayb
         /// </summary>
         public event Action OnGameplayStarted;
 
-        public override bool AllowBackButton => false; // handled by HoldForMenuButton
+        public override bool AllowExit => false; // handled by HoldForMenuButton
 
         protected override bool PlayExitSound => !isRestarting;
 
diff --git a/osu.Game/Screens/StartupScreen.cs b/osu.Game/Screens/StartupScreen.cs
index 9e04a238eb..ed0808c9ec 100644
--- a/osu.Game/Screens/StartupScreen.cs
+++ b/osu.Game/Screens/StartupScreen.cs
@@ -10,7 +10,7 @@ namespace osu.Game.Screens
     /// </summary>
     public abstract partial class StartupScreen : OsuScreen
     {
-        public override bool AllowBackButton => false;
+        public override bool AllowExit => false;
 
         public override bool HideOverlaysOnEnter => true;
 

```

Reply:
Sure, slightly baffled how this is happening in this PR but I'll keep that one to myself.

Reply:
Changes applied in https://github.com/ppy/osu/pull/30863/commits/3e1b4f4ac564a1b69b2d8111b19d3908f99980e4